File: Aman/Tree/Diameter of Binary Tree.cpp
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
       int h=0;
       return diameter(root,h);
    }
    int diameter(TreeNode* root, int& h)
    {
        if(root==NULL)
        {
            h=0;
            return 0;
        }
        int h1=0,h2=0;
        int d1=diameter(root->left,h1);
        int d2=diameter(root->right,h2);
        h=max(h1,h2)+1;
        return max({d1,d2,h1+h2});
    }
};

File: Rohit Singhal/Trees/DiameterofBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans=Integer.MIN_VALUE;
    public int dfs(TreeNode root)
    {
        if(root==null)
            return 0;
            
        int lh=dfs(root.left);
        int rh=dfs(root.right);
        if(lh+rh>ans)
            ans=lh+rh;
        
        return 1+Math.max(lh,rh);
    }
    
    
    public int diameterOfBinaryTree(TreeNode root) {
        int depth=dfs(root);
        return ans;
        
    }
}
